Libraries/Opc.Ua.Client/Subscription.cs

```diff
@@ -1819,42 +1842,81 @@ private async Task<(bool, UInt32Collection, UInt32Collection)> GetMonitoredItems
         private void StartKeepAliveTimer()
         {
             // stop the publish timer.
-            int keepAliveInterval;
             lock (m_cache)
             {
                 Utils.SilentDispose(m_publishTimer);
                 m_publishTimer = null;
 
-                m_lastNotificationTime = DateTime.UtcNow;
-                keepAliveInterval = (int)(Math.Min(m_currentPublishingInterval * m_currentKeepAliveCount, Int32.MaxValue));
-                m_publishTimer = new Timer(OnKeepAlive, keepAliveInterval, keepAliveInterval, keepAliveInterval);
+                Interlocked.Exchange(ref m_lastNotificationTime, DateTime.UtcNow.Ticks);
+                m_keepAliveInterval = (int)(Math.Min(m_currentPublishingInterval * (m_currentKeepAliveCount + 1), Int32.MaxValue));
+#if NET6_0_OR_GREATER
+                var publishTimer = new PeriodicTimer(TimeSpan.FromMilliseconds(m_keepAliveInterval));
+                Task.Run(() => OnKeepAliveAsync(publishTimer, CancellationToken.None));
+                m_publishTimer = publishTimer;
+#else
+                m_publishTimer = new Timer(OnKeepAlive, m_keepAliveInterval, m_keepAliveInterval, m_keepAliveInterval);
+#endif
 
                 if (m_messageWorkerTask == null || m_messageWorkerTask.IsCompleted)
                 {
-                    m_messageWorkerShutdownEvent.Reset();
-                    m_messageWorkerTask = Task.Run(() => PublishResponseMessageWorkerAsync());
+                    Utils.SilentDispose(m_messageWorkerCts);
+                    m_messageWorkerCts = new CancellationTokenSource();
+                    m_messageWorkerTask = Task.Run(() => PublishResponseMessageWorkerAsync(m_messageWorkerCts.Token));
                 }
             }
 
             // send initial publish.
-            m_session.BeginPublish(Math.Min(keepAliveInterval, Int32.MaxValue / 3) * 3);
+            m_session.BeginPublish(BeginPublishTimeout());
         }
 
+#if NET6_0_OR_GREATER
+        /// <summary>
+        /// Checks if a notification has arrived. Sends a publish if it has not.
+        /// </summary>
+        private async Task OnKeepAliveAsync(PeriodicTimer publishTimer, CancellationToken ct)
```

Comment:
Since (I believe) periodic timer is built on top of "TimerQueue" which is the same thing used by "Timer" you are not gaining anything here unless you are doing something "async" (like in the previous case where you read stuff). The only benefit is that the timer is "periodically" firing, i.e., if the HandleOnKeepAliveStopped takes longer than the period, the next period is firing.  Not sure though if that is actually intended.  I would stick with the existing "Timer" until we need to do something here that is "Async".

Reply:
periodic timer helps to avoid queued timer callbacks. Now the timer callback is non blocking and may be not much of a difference, but there is guarantee that there is only a single OnKeepAlive at a single time.